### PR TITLE
[fix](fe) Fix SimpleDateFormatter thread unsafe issue by replacing to DateTimeFormatter.

### LINKED ISSUE #19192: [Bug] SimpleDateFormat throws exceptions because it is not thread-safe.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

all versions

### What's Wrong?

```
2023-04-28 01:14:31,419 WARN (thrift-server-pool-4|245) [FrontendServiceImpl.streamLoadPut():1251] catch unknown result.
java.lang.ArrayIndexOutOfBoundsException: 960
        at sun.util.calendar.BaseCalendar.getCalendarDateFromFixedDate(BaseCalendar.java:453) ~[?:1.8.0_131]
        at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2397) ~[?:1.8.0_131]
        at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2312) ~[?:1.8.0_131]
        at java.util.Calendar.setTimeInMillis(Calendar.java:1804) ~[?:1.8.0_131]
        at java.util.Calendar.setTime(Calendar.java:1770) ~[?:1.8.0_131]
        at java.text.SimpleDateFormat.format(SimpleDateFormat.java:943) ~[?:1.8.0_131]
        at java.text.SimpleDateFormat.format(SimpleDateFormat.java:936) ~[?:1.8.0_131]
        at java.text.DateFormat.format(DateFormat.java:345) ~[?:1.8.0_131]
        at org.apache.doris.planner.StreamLoadPlanner.plan(StreamLoadPlanner.java:272) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.service.FrontendServiceImpl.streamLoadPutImpl(FrontendServiceImpl.java:1284) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.service.FrontendServiceImpl.streamLoadPut(FrontendServiceImpl.java:1245) ~[doris-fe.jar:1.2-SNAPSHOT]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
        at org.apache.doris.service.FeServer.lambda$start$0(FeServer.java:59) ~[doris-fe.jar:1.2-SNAPSHOT]
        at com.sun.proxy.$Proxy23.streamLoadPut(Unknown Source) ~[?:?]
        at org.apache.doris.thrift.FrontendService$Processor$streamLoadPut.getResult(FrontendService.java:2412) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.thrift.FrontendService$Processor$streamLoadPut.getResult(FrontendService.java:2392) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:38) ~[libthrift-0.16.0.jar:0.16.0]
        at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:38) ~[libthrift-0.16.0.jar:0.16.0]
        at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:250) ~[libthrift-0.16.0.jar:0.16.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]
```

### What You Expected?

No exception throws.

### How to Reproduce?

_No response_

### Anything Else?

We will use `DateTimeFormatter` to replace `SimpleDateFormat` to guarantee thread safe.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by CodeCooker17**
…teFormat is not thread-safe.

# Proposed changes

Issue Number: close #19192

## Problem summary

DateTimeFormatter replace SimpleDateFormat in fe module because SimpleDateFormat is not thread-safe.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java [hunk 1]
```diff
@@ -244,8 +246,8 @@ public void testDatePropInvalidString() {
         exceptionRule.expect(IllegalArgumentException.class);
         exceptionRule.expectMessage(Matchers.containsString("Invalid time format"));
 
-        PropertySchema.DateProperty prop = new PropertySchema.DateProperty("key", new SimpleDateFormat("yyyy-MM-dd "
-                + "HH:mm:ss"));
+        PropertySchema.DateProperty prop = new PropertySchema.DateProperty("key", DateTimeFormatter.ofPattern("yyyy-MM-dd "
+            + "HH:mm:ss").withZone(ZoneId.systemDefault()));
```
Comment (morningman) [inline fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:250]:
```
Error:  /home/runner/work/doris/doris/fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:250:13: '+' has incorrect indentation level 12, expected level should be 16. [Indentation]
Error:  /home/runner/work/doris/doris/fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:260:13: '+' has incorrect indentation level 12, expected level should be 16. [Indentation]
```

Comment (morningman) [reply fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:250]: code style

Comment (CodeCooker17) [reply fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:250]: @morningman ok, i will improve code style and solve these problems.

Comment (CodeCooker17) [reply fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java]: @morningman thanks your time, i have solve these problems.

**Comment by kaka11chen** [comment]
run buildall

**Comment by CodeCooker17** [comment]
>

**Comment by CodeCooker17** [comment]
>

**Comment by CodeCooker17** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.9 seconds
 stream load tsv:          423 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230508041535_clickbench_pr_139971.html

**Comment by CodeCooker17** [comment]
run buildall

**Comment by CodeCooker17** [comment]
run buildall

**Comment by kaka11chen** [comment]
LGTM.

File: fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java [hunk 2]
```diff
@@ -305,4 +314,11 @@ public static LocalDateTime formatDateTimeAndFullZero(String datetime, DateTimeF
                 LocalTime.of(hour, minute, second, milliSecond * 1000000));
     }
 
+    private static String formatDateStr(String dateStr) {
+        String[] parts = dateStr.trim().split("[ :-]+");
+        return String.format("%s-%02d-%02d%s", parts[0], Integer.parseInt(parts[1]), Integer.parseInt(parts[2]),
+                parts.length > 3 ? String.format(" %02d:%02d:%02d", Integer.parseInt(parts[3]),
+                        Integer.parseInt(parts[4]), Integer.parseInt(parts[5])) : "");
+    }
+
```
Comment (CodeCooker17) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:323]: SimpleDateFormat will try to parse it into a valid datetime according to certain rules when the incoming string does not exactly match the specified date format. For example, there is only one month in "2013-2-01", and SimpleDateFormat will format the string into "2013-02-01" for parsing. But DateTimeFormatter has strict format validation. So a formatting logic for date strings is added here to complete the month and day in the date format.

Comment (kaka11chen) [reply fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:323]: ok

File: fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java [hunk 3]
```diff
@@ -85,32 +84,42 @@ public class TimeUtils {
...
-            MAX_DATE = DATE_FORMAT.parse("9999-12-31");
 
-            MIN_DATETIME = DATETIME_FORMAT.parse("0000-01-01 00:00:00");
-            MAX_DATETIME = DATETIME_FORMAT.parse("9999-12-31 23:59:59");
+            MIN_DATE = Date.from(
+                    LocalDate.parse("0001-01-01", DATE_FORMAT).atStartOfDay().atZone(TIME_ZONE).toInstant());
+            MAX_DATE = Date.from(
+                    LocalDate.parse("9999-12-31", DATE_FORMAT).atStartOfDay().atZone(TIME_ZONE).toInstant());
+
+            MIN_DATETIME = Date.from(
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:117]: MIN_DATETIME is never used

File: fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java [hunk 4]
```diff
@@ -190,16 +197,18 @@ public static synchronized Date parseDate(String dateStr, PrimitiveType type) th
         if (!matcher.matches()) {
             throw new AnalysisException("Invalid date string: " + dateStr);
         }
+        dateStr = formatDateStr(dateStr);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:200]: Do we still need `synchronized` for this method?

File: fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java (hunk 3 above)
Comment (CodeCooker17) [reply fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:117]: MIN_DATE and MIN_DATETIME are only used in TimeUtilsTest

File: fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java (hunk 4 above)
Comment (CodeCooker17) [reply fe/fe-core/src/main/java/org/apache/doris/common/util/TimeUtils.java:200]: Methods using DateTimeFormatter do not need to be synchronized

File: fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java [hunk 5]
```diff
@@ -31,7 +31,9 @@
 import java.io.DataInputStream;
 import java.io.DataOutput;
 import java.io.DataOutputStream;
-import java.text.SimpleDateFormat;
```
[the remark this answers is not in the document]
Comment (CodeCooker17) [reply fe/fe-common/src/test/java/org/apache/doris/common/property/PropertySchemaTest.java:34]: added it

**Comment by CodeCooker17** [comment]
run buildall

**Comment by CodeCooker17** [comment]
run buildall

**Comment by CodeCooker17** [comment]
run buildall

**Review by morningman — COMMENTED** [review]
LGTM, please rebase to fix conflict

**Comment by CodeCooker17** [comment]
run buildall

**Comment by CodeCooker17** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Comment by qidaye** [comment]
Add error message for quick problem location.
```
2023-04-28 01:14:31,419 WARN (thrift-server-pool-4|245) [FrontendServiceImpl.streamLoadPut():1251] catch unknown result.
java.lang.ArrayIndexOutOfBoundsException: 960
        at sun.util.calendar.BaseCalendar.getCalendarDateFromFixedDate(BaseCalendar.java:453) ~[?:1.8.0_131]
        at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2397) ~[?:1.8.0_131]
        at java.util.GregorianCalendar.computeFields(GregorianCalendar.java:2312) ~[?:1.8.0_131]
        at java.util.Calendar.setTimeInMillis(Calendar.java:1804) ~[?:1.8.0_131]
        at java.util.Calendar.setTime(Calendar.java:1770) ~[?:1.8.0_131]
        at java.text.SimpleDateFormat.format(SimpleDateFormat.java:943) ~[?:1.8.0_131]
        at java.text.SimpleDateFormat.format(SimpleDateFormat.java:936) ~[?:1.8.0_131]
        at java.text.DateFormat.format(DateFormat.java:345) ~[?:1.8.0_131]
        at org.apache.doris.planner.StreamLoadPlanner.plan(StreamLoadPlanner.java:272) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.service.FrontendServiceImpl.streamLoadPutImpl(FrontendServiceImpl.java:1284) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.service.FrontendServiceImpl.streamLoadPut(FrontendServiceImpl.java:1245) ~[doris-fe.jar:1.2-SNAPSHOT]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
        at org.apache.doris.service.FeServer.lambda$start$0(FeServer.java:59) ~[doris-fe.jar:1.2-SNAPSHOT]
        at com.sun.proxy.$Proxy23.streamLoadPut(Unknown Source) ~[?:?]
        at org.apache.doris.thrift.FrontendService$Processor$streamLoadPut.getResult(FrontendService.java:2412) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.doris.thrift.FrontendService$Processor$streamLoadPut.getResult(FrontendService.java:2392) ~[doris-fe.jar:1.2-SNAPSHOT]
        at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:38) ~[libthrift-0.16.0.jar:0.16.0]
        at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:38) ~[libthrift-0.16.0.jar:0.16.0]
        at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:250) ~[libthrift-0.16.0.jar:0.16.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]
```

**Pull request merged**